Add Celery/Redis alternatives to GCP/pubsub

ericzinnikas opened the pull request:
Specifically, this uses [Celery](https://github.com/celery/celery) as an alternate for server/worker communication; [Kombu](https://github.com/celery/kombu) as an alternate for client/server communication (i.e. sending evidence); and redis as an alternative for the task state datastore.

Tested end-to-end processing of a sample rawdisk evidence locally and everything works as expected. Celery/Kombu support a bunch of different transports/backends, I used redis with them in my testing:

- spun up celery workers `./turbiniactl -C celeryworker`
- started the server `./turbiniactl -C -S server`
- sent evidence to process `./turbiniactl -o /mnt/turbinia/output/ -C rawdisk -l /Users/ericwz/SCHARDT.dd`
- queried status of tasks `./turbiniactl -C status`
- saw output csv: `less /mnt/turbinia/output/1527188930-5efe42d63b2c483687433bcfe41e8f74-PsortTask/5efe42d63b2c483687433bcfe41e8f74.csv`

I was unable to test GCP functionality, but didn't touch any of that code so see no reason why it shouldn't continue to function as before.

The only thing that probably still needs to get done is the splitting up of dependencies. I think we decided using `extras_require` and then having a `turbinia[celery]` target would be the way to do this.

aarontp commented on tests/client.py:
```diff
@@ -33,7 +33,8 @@ class TestTurbiniaClient(unittest.TestCase):
   def testTurbiniaClientInit(self, _):
     """Basic test for client."""
-    config.LoadConfig() client = TurbiniaClient()
+    config.LoadConfig()
```
That's.... Weird.  Don't know how that was submitted that way.  Might have happened when I was resolving merge commits somehow.

ericzinnikas commented on turbinia/client.py:
```diff
@@ -196,6 +196,44 @@ def send_request(self, request):
+      List of Task dict objects.
+    """
+    return self.redis.get_task_data(instance, days, task_id, request_id)
+
```
`turbiniactl status` still works fine, the call just passes down to the `wait_for_request()` in TurbiniaClient which then uses this version of `get_task_data()` (same with `format_task_status()` too).

ericzinnikas commented on turbinia/client.py:
```diff
@@ -196,6 +196,44 @@ def send_request(self, request):
+    self.task_manager.kombu.send_request(request)
+
+  def get_task_data(self, instance, _, __, days=0, task_id=None,
+                    request_id=None, function_name='gettasks'):
```
👍

ericzinnikas commented on turbinia/config/turbinia_config.py:
```diff
@@ -82,33 +82,57 @@
-INSTANCE = None
-DEVICE_NAME = None
-SCRATCH_PATH = None
+PROJECT = 'None'
```
+1

ericzinnikas commented on turbinia/pubsub.py:
```diff
@@ -94,6 +99,130 @@ def from_json(self, json_str):
+    )
+    self.app.conf.update(
+        task_default_queue=config.INSTANCE_ID,
+        event_serializer='pickle',
```
Seems you can do some Task class inheritance things: http://docs.celeryproject.org/en/latest/whatsnew-4.0.html#the-task-base-class-no-longer-automatically-register-tasks

So something like this might work `@app.task(bind=True, base=TurbiniaTask)`, but I will need to do more testing / figure that out (in a future PR probably).

ericzinnikas commented on turbinia/pubsub.py:
```diff
@@ -94,6 +99,130 @@ def from_json(self, json_str):
+  """Queue object for receiving evidence messages.
+
+  Attributes:
+    queue (Kombu.SimpleBuffer): evidence queue.
```
👍

ericzinnikas commented on turbinia/pubsub.py:
```diff
@@ -94,6 +99,130 @@ def from_json(self, json_str):
+    self.fexec = self._fexec()
+
+
+class TurbiniaKombu(object):
```
#185 👍

aarontp replied:
thanks

aarontp commented on turbinia/state_manager.py:
```diff
@@ -132,14 +138,83 @@ def update_task(self, task):
+      return
+    log.info('Updating task {0:s} in Datastore'.format(task.name))
+    task_data = self.get_task_dict(task)
+    task_data['last_update'] = task_data['last_update'].strftime(
```
Is this getting converted back and forth between datetime and string because of the json serialization?  Or does redis just not support datetime?  Can you maybe add a comment about this in either case?

aarontp commented on turbinia/task_manager.py:
```diff
@@ -251,6 +255,78 @@ def run(self):
+
+    outstanding_task_count = len(self.tasks) - len(completed_tasks)
+    log.info('{0:d} Tasks still outstanding.'.format(outstanding_task_count))
+    # pylint: disable=expression-not-assigned
```
I guess this can get removed.  I probably forgot to remove it in the PSQ object after changing the return not to be a list comprehension.

ericzinnikas commented on turbinia/pubsub.py:
```diff
@@ -94,6 +99,130 @@ def from_json(self, json_str):
     self.__dict__ = obj
 
 
+class TurbiniaCelery(object):
```
Yeah, maybe move both to celery.py (since they're dependent on one another)?

aarontp replied:
SGTM

ericzinnikas commented on turbinia/pubsub.py:
```diff
@@ -94,6 +99,130 @@ def from_json(self, json_str):
+        worker_concurrency=1,
+        worker_prefetch_multiplier=1,
+    )
+    self.fexec = self._fexec()
```
Agreed it is a bit weird, since `fexec` is set to the return value of the `_fexec` closure (which is another function). I'll work on updating the docstring and creating a placeholder method that documents the functionality.

aarontp replied:
thanks

ericzinnikas commented on turbinia/pubsub.py:
```diff
@@ -94,6 +99,130 @@ def from_json(self, json_str):
+    """Sends a TurbiniaRequest request"""
+    self.send_message(request.to_json())
+
+  def _validate_message(self, message):
```
Maybe a `TurbiniaMessage` base class? I can then move the common code down there and stub out the rest.

aarontp replied:
Yeah, I think that sounds good.  Thanks.

ericzinnikas commented:
Addressed most of your comments, still TODO:

- `pubsub.py` reorganization into multiple files (once I get feedback on my comment above)
- `requirements.txt` into extras (once your PR is decided, or should I just do it?)

aarontp commented:
I see you've pushed some changes since the last comment.  Let me know if/when this is ready for the re-review.  Thanks!

ericzinnikas commented:
We're good to go here for a re-review (sorry, end of half + FIRST travel keeping me occupied). Only real change (other than the shuffling around of classes/files) with this is the addition of a `TurbiniaMessageBase` class.  I also had to import `absolute_import` inside of `task_manager.py` to fix import ordering (`celery` module vs. `turbinia.celery` module).

Is it simplest if we have a separate PR to do all of the changes in 184 at once?

ericzinnikas commented:
Tested with `pip install .[local]` and `pip install .[worker]`. All seems to work as intended.
